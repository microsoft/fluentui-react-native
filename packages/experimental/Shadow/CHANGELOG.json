{
  "name": "@fluentui-react-native/experimental-shadow",
  "entries": [
    {
      "date": "Fri, 26 Aug 2022 18:15:16 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.13",
      "version": "0.2.13",
      "comments": {
        "patch": [
          {
            "author": "78454019+lyzhan7@users.noreply.github.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "661eeca6abbf496a2b36f2ecd4ae8dd6a6ae2f58",
            "comment": "Use undefined shadowToken in NotificationTokens"
          }
        ]
      }
    },
    {
      "date": "Mon, 15 Aug 2022 23:29:03 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.12",
      "version": "0.2.12",
      "comments": {
        "patch": [
          {
            "author": "78454019+lyzhan7@users.noreply.github.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "0f96f121a70581779472908ff4e00a914dcd1bf0",
            "comment": "Initial SPEC.md for shadows"
          }
        ]
      }
    },
    {
      "date": "Mon, 15 Aug 2022 22:12:03 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.11",
      "version": "0.2.11",
      "comments": {
        "patch": [
          {
            "author": "78454019+lyzhan7@users.noreply.github.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "5e9c0dfef3718735520706056cb1c5f62524efd7",
            "comment": "Pass key shadow props into child instead of a new view"
          }
        ]
      }
    },
    {
      "date": "Mon, 25 Jul 2022 21:27:20 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.10",
      "version": "0.2.10",
      "comments": {
        "patch": [
          {
            "author": "sanajmi@microsoft.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "0772c5674928d890bf843ca0891f646f58ff0673",
            "comment": "Combine platform specific test apps"
          }
        ]
      }
    },
    {
      "date": "Wed, 20 Jul 2022 16:24:28 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.9",
      "version": "0.2.9",
      "comments": {
        "patch": [
          {
            "author": "ruaraki@microsoft.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "1de042e4c2b76d468203365a38f4ab4eeb95cc77",
            "comment": "Fix version of theme-types to align with other packages"
          }
        ]
      }
    },
    {
      "date": "Tue, 19 Jul 2022 18:10:02 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.8",
      "version": "0.2.8",
      "comments": {
        "patch": [
          {
            "author": "sanajmi@microsoft.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "1b9087c5526b378e6b35d70cf44d109c1b0dad5e",
            "comment": "Empty change"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/experimental-shadow",
            "comment": "Bump @fluentui-react-native/theme-types to v0.18.0",
            "commit": "1b9087c5526b378e6b35d70cf44d109c1b0dad5e"
          }
        ]
      }
    },
    {
      "date": "Fri, 15 Jul 2022 01:22:54 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.7",
      "version": "0.2.7",
      "comments": {
        "patch": [
          {
            "author": "78454019+lyzhan7@users.noreply.github.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "b3e63050f4b326ea7326a7d8e5c956537e9f5141",
            "comment": "Replace shadowDepth with shadowToken"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/experimental-shadow",
            "comment": "Bump @fluentui-react-native/theme-types to v0.18.0",
            "commit": "4a2991bd1d5bbcc38cfb023c84845557f4791de9"
          }
        ]
      }
    },
    {
      "date": "Fri, 15 Jul 2022 00:29:27 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.6",
      "version": "0.2.6",
      "comments": {
        "patch": [
          {
            "author": "ruaraki@microsoft.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "32c99c628be487a61c9ad2f792accc4360e1cfe7",
            "comment": "Fix some versions"
          },
          {
            "author": "78454019+lyzhan7@users.noreply.github.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "368fee7d4df7596cab2638938835f7ded84bc6c9",
            "comment": "Add shadow to iOS FAB"
          }
        ]
      }
    },
    {
      "date": "Thu, 14 Jul 2022 18:09:51 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.5",
      "version": "0.2.5",
      "comments": {
        "patch": [
          {
            "author": "duomaster01@gmail.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "155145b496a51b06bba73f8f9c6b89dc7b348c4b",
            "comment": "updated to rn 0.66"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/experimental-shadow",
            "comment": "Bump @fluentui-react-native/framework to v0.8.0",
            "commit": "155145b496a51b06bba73f8f9c6b89dc7b348c4b"
          }
        ]
      }
    },
    {
      "date": "Thu, 14 Jul 2022 16:13:45 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.4",
      "version": "0.2.4",
      "comments": {
        "patch": [
          {
            "author": "ruaraki@microsoft.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "3066715e32b200e751633c1979f5f2155fbd42fa",
            "comment": "Revert \"Add shadows to iOS FAB (#1839)\""
          }
        ]
      }
    },
    {
      "date": "Wed, 13 Jul 2022 23:23:59 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.3",
      "version": "0.2.3",
      "comments": {
        "patch": [
          {
            "author": "78454019+lyzhan7@users.noreply.github.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "d4320c9f7385a7f37dc8ce521c413071b8cb55a0",
            "comment": "Add shadow to FAB"
          }
        ]
      }
    },
    {
      "date": "Fri, 08 Jul 2022 21:23:37 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.2",
      "version": "0.2.2",
      "comments": {
        "patch": [
          {
            "author": "beachball",
            "package": "@fluentui-react-native/experimental-shadow",
            "comment": "Bump @fluentui-react-native/framework to v0.7.32",
            "commit": "c489bb685c8dbe94b7b6fe88957e7ce4d3c2d000"
          }
        ]
      }
    },
    {
      "date": "Fri, 08 Jul 2022 18:26:51 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.1",
      "version": "0.2.1",
      "comments": {
        "patch": [
          {
            "author": "78454019+lyzhan7@users.noreply.github.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "ef2a18b53bbc59178c30d1250ff89e64bb2ca5be",
            "comment": "Update + move shadowStyleFromTheme method"
          }
        ]
      }
    },
    {
      "date": "Tue, 28 Jun 2022 00:53:43 GMT",
      "tag": "@fluentui-react-native/experimental-shadow_v0.2.0",
      "version": "0.2.0",
      "comments": {
        "minor": [
          {
            "author": "78454019+lyzhan7@users.noreply.github.com",
            "package": "@fluentui-react-native/experimental-shadow",
            "commit": "b2d247cc39d7d3c8ca426f362a05f04e06b54d9b",
            "comment": "Rename shadow to experimental-shadow"
          }
        ]
      }
    }
  ]
}
