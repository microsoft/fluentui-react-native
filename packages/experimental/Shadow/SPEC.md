# Shadow

## Background

A `Shadow` is a component that can be used to add Fluent shadows to other components.

Shadows defined in the Fluent design system are actually comprised of two shadows. The **key shadow** is used to create a feeling of distance, and the **ambient shadow** is used to define the edges of the shape.

The `Shadow` component is different from other components in that it was created to address limitations with React Native - Views in React Native can only have one Shadow set at a time. This component provides an easy way to add a Fluent shadow to another component.

## Shadow Depth Ramp

These are the current `Shadow` variants:

|                 | Light mode                                                                                                                                      | Dark mode                                                                                                                                      |
| --------------- | ----------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |
| Regular shadows | <img width="820" alt="light" src="https://user-images.githubusercontent.com/78454019/187567525-64592ead-c64c-4cc1-acd3-a21abdea2d35.png">       | <img width="820" alt="dark" src="https://user-images.githubusercontent.com/78454019/187567552-3b376f4d-eacd-482d-a395-a261c374c9f6.png">       |
| Brand shadows   | <img width="820" alt="light-brand" src="https://user-images.githubusercontent.com/78454019/187567541-c939ceaa-35b1-4cd6-84b5-55d8a5032e0c.png"> | <img width="820" alt="dark-brand" src="https://user-images.githubusercontent.com/78454019/187567558-79783f9b-2bff-463c-b7b5-7ec692b000fd.png"> |

## Sample Code

Examples adding some Shadows to a Button:

```tsx
<Shadow shadowToken={theme.shadows.shadow8}>
  <Button>Text box with shadow8</Button>
</Shadow>
```

```tsx
<Shadow shadowToken={theme.shadows.shadow8brand}>
  <Button>Text box with shadow8brand</Button>
</Shadow>
```

For more examples of using Shadow, please see the [ShadowTest test page](https://github.com/microsoft/fluentui-react-native/tree/main/apps/fluent-tester/src/TestComponents/Shadow) in the [Fluent Tester app](https://github.com/microsoft/fluentui-react-native/blob/main/apps/fluent-tester/README.md).

For an example of adding a Shadow as a slot to a Fluent component, please see the [FAB component](https://github.com/microsoft/fluentui-react-native/tree/main/packages/components/Button/src/FAB) - this component exists on both iOS and Android, but currently only the iOS version uses the Shadow component. The [Notification component](https://github.com/microsoft/fluentui-react-native/tree/main/packages/components/Notification) is another example that uses the Shadow component.

## API

### Props

The `Shadow` component takes in one prop called `ShadowToken`. The variants of ShadowToken are all defined in the theme.
If `undefined` is passed into the `shadowToken` prop, no shadow will be rendered.

`Shadow` must take exactly one child, which must be of type View.

## Implementation details

In order to get around the React Native restriction of one Shadow per View, the Shadow component adds an extra View around the original View. One of the shadows is placed on the original View and the other is placed on the extra View. The Shadow component smartly extracts the original View's style props so that the extra View isn't noticable (ex. will always be the same size as the original View, padding/margins are set correctly, etc.)

## Notes

- Known issue: The Shadow component does not currently work if you are wrapping an Animated component.
- Known issue: there may be some slight rounding discrepancies due to hex to decimal rounding errors, ex. opacity 0.24 may end up at 0.25.
- One additional adjustment that was needed on Apple platforms was to divide the blur radius token value by 2. This was needed because shadow blur/radius looks different across platforms - more information here: https://github.com/microsoft/apple-ux-guide/blob/gh-pages/Shadows.md
