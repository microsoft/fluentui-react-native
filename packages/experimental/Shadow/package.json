{
  "name": "@fluentui-react-native/experimental-shadow",
  "version": "0.2.62",
  "description": "A cross-platform Shadow component using the Fluent Design System",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "typings": "lib/index.d.ts",
  "onPublish": {
    "main": "lib-commonjs/index.js",
    "module": "lib/index.js"
  },
  "scripts": {
    "build": "fluentui-scripts build",
    "clean": "fluentui-scripts clean",
    "depcheck": "fluentui-scripts depcheck",
    "just": "fluentui-scripts",
    "lint": "fluentui-scripts eslint",
    "test": "fluentui-scripts jest",
    "update-snapshots": "fluentui-scripts jest -u",
    "prettier": "fluentui-scripts prettier",
    "prettier-fix": "fluentui-scripts prettier --fix true"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/microsoft/fluentui-react-native.git",
    "directory": "packages/experimental/Shadow"
  },
  "dependencies": {
    "@fluentui-react-native/framework": "0.8.36",
    "@fluentui-react-native/pressable": "0.9.52",
    "@fluentui-react-native/theme-types": "0.30.0"
  },
  "devDependencies": {
    "@fluentui-react-native/eslint-config-rules": "^0.1.1",
    "@fluentui-react-native/scripts": "^0.1.1",
    "@fluentui-react-native/test-tools": ">=0.1.1 <1.0.0",
    "react": "17.0.2",
    "react-native": "^0.68.0"
  },
  "peerDependencies": {
    "react": "17.0.2",
    "react-native": "^0.68.0"
  },
  "author": "",
  "license": "MIT",
  "rnx-kit": {
    "reactNativeVersion": "^0.68",
    "reactNativeDevVersion": "^0.68",
    "kitType": "library",
    "capabilities": [
      "core",
      "core-android",
      "core-ios",
      "react"
    ]
  }
}
