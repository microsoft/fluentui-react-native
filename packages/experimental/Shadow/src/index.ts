export { Shadow } from './Shadow';
export type { ShadowProps } from './Shadow.types';
export { getShadowTokenStyleSet } from './shadowStyle';
