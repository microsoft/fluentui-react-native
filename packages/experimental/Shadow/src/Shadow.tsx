import * as React from 'react';
import { View } from 'react-native';
import { ShadowProps, shadowName } from './Shadow.types';
import { mergeProps, stagedComponent } from '@fluentui-react-native/framework';
import { getShadowTokenStyleSet } from './shadowStyle';

export const Shadow = stagedComponent((props: ShadowProps) => {
  return (final: ShadowProps, children: React.ReactNode) => {
    if (!props.shadowToken) {
      return <>{children}</>;
    }

    const shadowTokenStyleSet = getShadowTokenStyleSet(props.shadowToken);
    const mergedProps = mergeProps(final, { style: shadowTokenStyleSet.ambient });

    const childrenArray = React.Children.toArray(children) as React.ReactElement[];
    const child = childrenArray[0];

    if (__DEV__) {
      if (childrenArray.length !== 1) {
        console.warn('Shadow must only have one child');
      }
    }

    const childWithKeyShadow = React.cloneElement(child, mergeProps(child.props, { style: shadowTokenStyleSet.key }));

    return <View {...mergedProps}>{childWithKeyShadow}</View>;
  };
});

Shadow.displayName = shadowName;

export default Shadow;
