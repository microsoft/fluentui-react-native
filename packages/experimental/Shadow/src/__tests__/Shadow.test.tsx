import * as React from 'react';
import { Text, View } from 'react-native';

import { mergeStyles, useFluentTheme } from '@fluentui-react-native/framework';
import { Pressable } from '@fluentui-react-native/pressable';
import { checkRenderConsistency, checkReRender } from '@fluentui-react-native/test-tools';
import * as renderer from 'react-test-renderer';

import { Shadow } from '../Shadow';

const backgroundColor = { backgroundColor: 'red' };
interface ShadowTestProps {
  displayText: string;
  depth: string;
}

const TestShadow: React.FunctionComponent<ShadowTestProps> = (props: ShadowTestProps) => {
  const theme = useFluentTheme();
  return (
    <Shadow shadowToken={theme.shadows[props.depth]}>
      <View style={backgroundColor}>
        <Text>{props.displayText}</Text>
      </View>
    </Shadow>
  );
};

const TestPressableWithAndWithoutShadow: React.FunctionComponent = () => {
  const theme = useFluentTheme();
  return (
    <View>
      <Shadow shadowToken={theme.shadows['shadow16']}>
        <Pressable style={backgroundColor} />
      </Shadow>
      <View>
        <Pressable style={backgroundColor} />
      </View>
    </View>
  );
};

interface ShadowOnChildViewWithProps {
  childViewStyleProps: object;
}

const TestShadowOnChildViewWithProps: React.FunctionComponent<ShadowOnChildViewWithProps> = (props: ShadowOnChildViewWithProps) => {
  const theme = useFluentTheme();
  return (
    <Shadow shadowToken={theme.shadows['shadow16']}>
      <View style={mergeStyles(props.childViewStyleProps, backgroundColor)}>
        <Text>{JSON.stringify(props.childViewStyleProps)}</Text>
      </View>
    </Shadow>
  );
};

describe('Shadow component tests', () => {
  beforeAll(() => {
    jest.mock('react-native/Libraries/Utilities/Platform', () => ({
      OS: 'macos',
      select: () => null,
    }));
  });

  it('Shadow (depth=2)', () => {
    const tree = renderer.create(<TestShadow displayText="Shadow (depth=2)" depth="shadow2" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shadow (depth=4)', () => {
    const tree = renderer.create(<TestShadow displayText="Shadow (depth=4)" depth="shadow4" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shadow (depth=8)', () => {
    const tree = renderer.create(<TestShadow displayText="Shadow (depth=8)" depth="shadow8" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shadow (depth=16)', () => {
    const tree = renderer.create(<TestShadow displayText="Shadow (depth=16)" depth="shadow16" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shadow (depth=28)', () => {
    const tree = renderer.create(<TestShadow displayText="Shadow (depth=28)" depth="shadow28" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shadow (depth=64)', () => {
    const tree = renderer.create(<TestShadow displayText="Shadow (depth=64)" depth="shadow64" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Brand shadow (depth=2)', () => {
    const tree = renderer.create(<TestShadow displayText="Brand shadow (depth=2)" depth="shadow2brand" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Brand shadow (depth=4)', () => {
    const tree = renderer.create(<TestShadow displayText="Brand shadow (depth=4)" depth="shadow4brand" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Brand shadow (depth=8)', () => {
    const tree = renderer.create(<TestShadow displayText="Brand shadow (depth=8)" depth="shadow8brand" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Brand shadow (depth=16)', () => {
    const tree = renderer.create(<TestShadow displayText="Brand shadow (depth=16)" depth="shadow16brand" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Brand shadow (depth=28)', () => {
    const tree = renderer.create(<TestShadow displayText="Brand shadow (depth=28)" depth="shadow28brand" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Brand shadow (depth=64)', () => {
    const tree = renderer.create(<TestShadow displayText="Brand shadow (depth=64)" depth="shadow64brand" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Pressable that has a shadow vs. pressable without shadow', () => {
    const tree = renderer.create(<TestPressableWithAndWithoutShadow />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shadow on a child with margin and padding', () => {
    const tree = renderer.create(<TestShadowOnChildViewWithProps childViewStyleProps={{ margin: 2, padding: 2 }} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shadow on a child with border radius', () => {
    const tree = renderer.create(<TestShadowOnChildViewWithProps childViewStyleProps={{ borderRadius: 2 }} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shadow on a child with border width', () => {
    const tree = renderer.create(<TestShadowOnChildViewWithProps childViewStyleProps={{ borderWidth: 2 }} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shadow simple rendering does not invalidate styling', () => {
    checkRenderConsistency(() => <TestShadow displayText="Shadow render test" depth="shadow2" />, 2);
  });

  it('Shadow re-renders correctly', () => {
    checkReRender(() => <TestShadow displayText="Shadow render twice test" depth="shadow2" />, 2);
  });

  afterAll(() => {
    jest.unmock('react-native/Libraries/Utilities/Platform');
  });
});
