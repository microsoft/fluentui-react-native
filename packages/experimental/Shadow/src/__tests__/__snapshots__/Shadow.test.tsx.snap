// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shadow component tests Brand shadow (depth=2) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.3,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 1,
      }
    }
  >
    <Text>
      Brand shadow (depth=2)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Brand shadow (depth=4) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.3,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 2,
      }
    }
  >
    <Text>
      Brand shadow (depth=4)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Brand shadow (depth=8) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.3,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 4,
      }
    }
  >
    <Text>
      Brand shadow (depth=8)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Brand shadow (depth=16) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.3,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 8,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 8,
      }
    }
  >
    <Text>
      Brand shadow (depth=16)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Brand shadow (depth=28) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.3,
      "shadowRadius": 4,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 14,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 14,
      }
    }
  >
    <Text>
      Brand shadow (depth=28)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Brand shadow (depth=64) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.3,
      "shadowRadius": 4,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 32,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 32,
      }
    }
  >
    <Text>
      Brand shadow (depth=64)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Pressable that has a shadow vs. pressable without shadow 1`] = `
<View>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0.12,
        "shadowRadius": 1,
      }
    }
  >
    <View
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "red",
          "shadowColor": "#000000",
          "shadowOffset": {
            "height": 8,
            "width": 0,
          },
          "shadowOpacity": 0.14,
          "shadowRadius": 8,
        }
      }
    />
  </View>
  <View>
    <View
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "red",
        }
      }
    />
  </View>
</View>
`;

exports[`Shadow component tests Shadow (depth=2) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.12,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.14,
        "shadowRadius": 1,
      }
    }
  >
    <Text>
      Shadow (depth=2)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Shadow (depth=4) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.12,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.14,
        "shadowRadius": 2,
      }
    }
  >
    <Text>
      Shadow (depth=4)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Shadow (depth=8) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.12,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.14,
        "shadowRadius": 4,
      }
    }
  >
    <Text>
      Shadow (depth=8)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Shadow (depth=16) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.12,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 8,
          "width": 0,
        },
        "shadowOpacity": 0.14,
        "shadowRadius": 8,
      }
    }
  >
    <Text>
      Shadow (depth=16)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Shadow (depth=28) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.2,
      "shadowRadius": 4,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 14,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 14,
      }
    }
  >
    <Text>
      Shadow (depth=28)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Shadow (depth=64) 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.2,
      "shadowRadius": 4,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 32,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 32,
      }
    }
  >
    <Text>
      Shadow (depth=64)
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Shadow on a child with border radius 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "borderRadius": 2,
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.12,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "borderRadius": 2,
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 8,
          "width": 0,
        },
        "shadowOpacity": 0.14,
        "shadowRadius": 8,
      }
    }
  >
    <Text>
      {"borderRadius":2}
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Shadow on a child with border width 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.12,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "borderWidth": 2,
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 8,
          "width": 0,
        },
        "shadowOpacity": 0.14,
        "shadowRadius": 8,
      }
    }
  >
    <Text>
      {"borderWidth":2}
    </Text>
  </View>
</View>
`;

exports[`Shadow component tests Shadow on a child with margin and padding 1`] = `
<View
  style={
    {
      "backgroundColor": "red",
      "margin": 2,
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.12,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "red",
        "padding": 2,
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 8,
          "width": 0,
        },
        "shadowOpacity": 0.14,
        "shadowRadius": 8,
      }
    }
  >
    <Text>
      {"margin":2,"padding":2}
    </Text>
  </View>
</View>
`;
